src/Covenant/Data.hs

```diff
@@ -143,3 +177,45 @@ mkBaseFunctor (DataDeclaration tn numVars ctors) = do
     -- This tells us whether the ValT *is* a recursive child of the parent type
     isRecursive :: ValT AbstractTy -> Reader ScopeBoundary Bool
     isRecursive = isRecursiveChildOf tn
+
+-- data Maybe a = Just a | Nothing ~> forall out. (a -> out) -> out -> out
+
+-- Only returns `Nothing` if there are no Constructors
+mkBBF :: DataDeclaration AbstractTy -> Maybe (ValT AbstractTy)
+mkBBF (DataDeclaration _ numVars ctors)
+  | V.null ctors = Nothing
+  | otherwise = ThunkT . CompT bbfCount . CompTBody . flip NEV.snoc topLevelOut <$> (NEV.fromVector =<< traverse mkEliminator ctors)
+  where
+    mkEliminator :: Constructor AbstractTy -> Maybe (ValT AbstractTy)
+    mkEliminator (Constructor _ (fmap incAbstractionDB -> args))
+      | V.null args = Just topLevelOut
+      | otherwise =
+          let out = Abstraction $ BoundAt (S Z) outIx
+           in ThunkT . CompT count0 . CompTBody . flip NEV.snoc out <$> NEV.fromVector args
+
+    incAbstractionDB :: ValT AbstractTy -> ValT AbstractTy
+    incAbstractionDB = mapValT $ \case
+      Abstraction (BoundAt db indx) ->
+        let db' = fromJust . preview asInt $ review asInt db + 1
+         in Abstraction (BoundAt db' indx)
+      other -> other
+
+    topLevelOut = Abstraction $ BoundAt Z outIx
+
+    -- We have to add a type parameter for the result.
+    bbfCount = fromJust . preview intCount $ review intCount numVars + 1
+
+    -- The index of the 'out' parameter, indicates the return type
+    outIx :: Index "tyvar"
+    outIx = fromJust . preview intIndex $ review intCount numVars
+
+{- Note (largely to self) on DeBruijn indices:
```

Comment:
I've been using a scheme for notes which does something like

-- Note (Koz, 15/05/25): Buy milk.

Might as well be consistent.